In [1]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
import statsmodels.formula.api as smf
import matplotlib.pyplot as plt
from scipy import stats
import statsmodels.stats.api as sms

Usando el dataset de Casas, genere esta misma regresión usando Excel, Stata y Statsmodels.
1) Interprete el coeficiente 4139.8 de la variable GARAGE.
2) ¿Para qué sirve el estadístico F (=96.14), cómo está calculado y a qué conclusión puede llegar con el mismo? Indique claramente de qué hipótesis nula y alternativa es este estadístico. Es decir, explique cuál sería el modelo bajo las hipótesis nula y alternativa.
3) ¿Puede haber evidencia de multicolinealidad en la Regresión 1? Justifique claramente.
4) ¿Cuál es el valor esperado de una casa en un lote de 5100 pies cuadrados, con 3 cuartos, 2 baños, construida en dos pisos, con entrada para auto, cuarto de recreación, calefacción central en un vecindario agradable (NBHD=1), sin garage, ni sótano, ni aire acondicionado central? Muestre cómo obtiene ese valor.

In [4]:
# Cargo el dataset de casas

dfcasa = pd.read_excel('Ejemplo_Casa.xls', sheet_name='HPRICE', usecols='A:L')
dfcasa = dfcasa[dfcasa["BANOS"] != 4]

dfcasa.head()

PRECIO  LOTE  CUARTOS  BANOS  PISOS  ENTRADA  REC  SOTANO  CALEF  AIRE  \
0   42000  5850        3      1      2        1    0       1      0     0   
1   38500  4000        2      1      1        1    0       0      0     0   
2   49500  3060        3      1      1        1    0       0      0     0   
3   60500  6650        3      1      2        1    1       0      0     0   
4   61000  6360        2      1      1        1    0       0      0     0   

   GARAGE  NBHD  
0       1     0  
1       0     0  
2       0     0  
3       0     0  
4       0     0

In [5]:
y = dfcasa['PRECIO'] # target
X = dfcasa.drop(columns='PRECIO')
X = sm.add_constant(X) # añadimos la constante 

X.head()

const  LOTE  CUARTOS  BANOS  PISOS  ENTRADA  REC  SOTANO  CALEF  AIRE  \
0    1.0  5850        3      1      2        1    0       1      0     0   
1    1.0  4000        2      1      1        1    0       0      0     0   
2    1.0  3060        3      1      1        1    0       0      0     0   
3    1.0  6650        3      1      2        1    1       0      0     0   
4    1.0  6360        2      1      1        1    0       0      0     0   

   GARAGE  NBHD  
0       1     0  
1       0     0  
2       0     0  
3       0     0  
4       0     0

In [6]:
regmul = sm.OLS(y, X).fit()
regmul.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 PRECIO   R-squared:                       0.665
Model:                            OLS   Adj. R-squared:                  0.658
Method:                 Least Squares   F-statistic:                     96.14
Date:                Fri, 01 Aug 2025   Prob (F-statistic):          7.67e-119
Time:                        16:05:52   Log-Likelihood:                -6022.2
No. Observations:                 545   AIC:                         1.207e+04
Df Residuals:                     533   BIC:                         1.212e+04
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -3509.7393   3419.390     -1.026      0.305   -1.02e+04    3207.395
LOTE           3.5341      0.350     10.102      0.000       2.847       4.221
CUARTOS     1909.3244   1046.422      1.825      0.069    -146.293    3964.941
BANOS       1.383e+04   1519.347      9.102      0.000    1.08e+04    1.68e+04
PISOS       6492.6879    924.670      7.022      0.000    4676.244    8309.132
ENTRADA     6722.7877   2042.146      3.292      0.001    2711.146    1.07e+04
REC         4623.2425   1898.206      2.436      0.015     894.360    8352.125
SOTANO      5506.6270   1585.876      3.472      0.001    2391.293    8621.961
CALEF         1.3e+04   3214.103      4.043      0.000    6681.465    1.93e+04
AIRE        1.262e+04   1552.591      8.130      0.000    9572.753    1.57e+04
GARAGE      4139.8173    841.671      4.919      0.000    2486.418    5793.216
NBHD        9469.9610   1667.598      5.679      0.000    6194.090    1.27e+04
==============================================================================
Omnibus:                       97.367   Durbin-Watson:                   1.595
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              263.699
Skew:                           0.881   Prob(JB):                     5.48e-58
Kurtosis:                       5.917   Cond. No.                     3.08e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.08e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

1) Que garage sea 4139, significa que el precio aumenta en 4139 unidades (manteniendo todas las demas constantes, ceteris paribus)

2) Este estadístico se usa para testear la hipótesis nula de que todas las pendientes son iguales a cero simultáneamente contra la hipótesis alternativa de que al menos una de estas pendientes es distinta de cero.

3) para que haya multicolinealidad entre dos o más variables, los p-values deberían ser mayores a 0,10. Esto ocurre ya que cuando hay multicolinealidad, los errores estándar de las pendientes (de los beta sombrero) que ultiplican a las variables explicativas son colineales entre sí, tendrán valores altos, repercutiendo en un estadístico t bajo y un p-value alto. Esto no ocurre en esta regresión, de manera que podemos descartar ulticolinealidad. Pero si ese hubiera sido el caso, uno debe eliminar las variables involucradas de a una y fijarse si la(s) que queda(n) pasan a tener p-values bajos. Alternativamente uno puede correr un test de hipótesis de que los betas que multiplican a todas esas variables explicativas que parecieran no explicar a la variable dependiente son simultáneamente iguales a cero. Si rechazo esta hipótesis nula, entonces tengo un problema de multicolinealidad.

4) Hay que multiplicar las constantes por los valores dados y ese seria el precio

Correr la misma regresion pero ahora se elimina la variable BANOS y se incorporan las variables DB2 y DB3.
1) ¿Cómo se interpreta el coeficiente 29392.54 correspondiente a la variable DB3? Indique claramente.
2) Obtenga un intervalo de confianza del 95% para el beta que multiplica a la variable lote a partir de la información de la segunda regresión (REGRESION 2). Puede usar la distribución normal como aproximación. También puede responder usando algún software, en cuyo caso deberá indicar claramente qué hizo.
3) Suponga que le interesa testear con una probabilidad de error tipo I de 5% si el beta que multiplica a DB3 es el doble al beta que multiplica a DB2. Es decir, desea testear si βDB3 = 2βDB2. ¿Puede realizar este test de hipótesis? ¿Tiene alguna intuición testear particularmente esta hipótesis, βDB3 = 2βDB2? ¿Qué significa intuitivamente esta hipótesis nula, es decir, qué es lo que se pretende testear con este test de hipótesis? Explique claramente qué pasos realizó en el software, cómo obtuvo el estadístico (su valor), el p-value y la conclusión a la que llega. Explique cada paso en detalle

In [7]:
regsmf = smf.ols('PRECIO ~ LOTE + CUARTOS + C(BANOS) + PISOS + ENTRADA + REC + SOTANO + CALEF + AIRE + GARAGE + NBHD', data=dfcasa).fit()
regsmf.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 PRECIO   R-squared:                       0.665
Model:                            OLS   Adj. R-squared:                  0.657
Method:                 Least Squares   F-statistic:                     88.01
Date:                Fri, 01 Aug 2025   Prob (F-statistic):          7.04e-118
Time:                        16:45:55   Log-Likelihood:                -6022.1
No. Observations:                 545   AIC:                         1.207e+04
Df Residuals:                     532   BIC:                         1.213e+04
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept      1.031e+04   3464.573      2.976      0.003    3503.032    1.71e+04
C(BANOS)[T.2]  1.348e+04   1726.361      7.809      0.000    1.01e+04    1.69e+04
C(BANOS)[T.3]  2.939e+04   5079.824      5.786      0.000    1.94e+04    3.94e+04
LOTE              3.5268      0.351     10.062      0.000       2.838       4.215
CUARTOS        1902.8737   1047.336      1.817      0.070    -154.547    3960.295
PISOS          6527.3156    928.943      7.027      0.000    4702.469    8352.162
ENTRADA        6747.2925   2044.525      3.300      0.001    2730.960    1.08e+04
REC            4623.6067   1899.665      2.434      0.015     891.842    8355.371
SOTANO         5521.0982   1587.458      3.478      0.001    2402.642    8639.554
CALEF          1.302e+04   3217.108      4.047      0.000    6700.508    1.93e+04
AIRE            1.27e+04   1563.394      8.121      0.000    9625.256    1.58e+04
GARAGE         4141.7535    842.330      4.917      0.000    2487.052    5796.455
NBHD           9496.8991   1670.077      5.687      0.000    6216.144    1.28e+04
==============================================================================
Omnibus:                       97.932   Durbin-Watson:                   1.597
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              268.530
Skew:                           0.882   Prob(JB):                     4.89e-59
Kurtosis:                       5.952   Cond. No.                     4.35e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.35e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

1) Esto significa cuanto aumenta el precio de una casa al tener 3 baños comparado a una casa que posea 1 solo baño, manteniendo todas las demas variables fijas.

2) Que nos da: 𝛽̂ ± 1.96 ∗ 𝑠𝑒(𝛽̂) , de modo que el intervalo de confianza del 95% para el beta que premultiplica a LOTE es 3.536 ± 1.96 ∗ 0.35, o sea, 𝛽 ∈ (2.85 , 4.22).

In [ ]:
regsmf.conf_int(alpha=0.05)

""" 
El IC del 95% para Lote es [2.84 ; 4.22]
"""

0             1
Intercept       3503.031922  17114.874638
C(BANOS)[T.2]  10089.861601  16872.502352
C(BANOS)[T.3]  19413.565360  39371.514991
LOTE               2.838182      4.215325
CUARTOS         -154.547238   3960.294637
PISOS           4702.468972   8352.162219
ENTRADA         2730.960461  10763.624507
REC              891.841877   8355.371432
SOTANO          2402.642322   8639.554024
CALEF           6700.507506  19340.094078
AIRE            9625.255576  15767.623194
GARAGE          2487.052225   5796.454715
NBHD            6216.144015  12777.654123

3) Se puede responder usando los RSS de la regresión 2 (que sería el URSS) y usando el RSS de la regresión 1 (que sería el RRSS). El RSS de la regresión 1 es el RSS del modelo restringido ya que BANOS = 1 + DB2 + 2*DB3, de modo que imponer la hipótesis nula, βDB3 = 2βDB2, quedaría un modelo que al estimarlo nos va a dar el mismo RSS que el de la regresión 1. El valor del estadístico F (notar que q=1, n=545, k=13) se obtiene haciendo el cálculo del mismo. El estadístico F tiene entonces 1 grado de libertad en el numerador y 532 grados de libertad en el denominador y se calcula usando la fórmula 
𝐹 = (𝑅𝑅𝑆𝑆 − 𝑈𝑅𝑆𝑆)/𝑞 / 𝑈𝑅𝑆𝑆/(𝑛 − 𝑘)
El estadístico F da un valor de 0.1825 y el p-value asociado es 0.67 (este valor se obtiene si usan un software), y consecuentemente no rechazamos la hipótesis nula.

In [9]:
print(regsmf.f_test("C(BANOS)[T.3] = 2 * C(BANOS)[T.2]"))

<F test: F=0.18151291728567842, p=0.6702485908497321, df_denom=532, df_num=1>
